### ISSUE TITLE
Uncaught Error in Assignment of type Iterator[int] to type Iterator[str]

### ISSUE THREAD

**Issue by didi64** [python/mypy#19675]
**Bug Report**

Mypy allows the assignment 
it: Iterator[str] = iter(d) where d is dict[int, str].

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=05c55a654acd175b763be1c73567a107

```python
from typing import Iterator


d: dict[int, str] = {1: 'A'}
it: Iterator[str] = iter(d)  # mypy does not complain, although iter(d) has type Iterator[int]!
s = next(it) + 'foo'  # TypeError: unsupported operand type(s) for +: 'int' and 'str'
```

**Expected Behavior**

Mypy raises an error such as
error: Argument 1 to "iter" has incompatible type "dict[int, str]"; expected "SupportsIter[Iterator[str]]"

**Actual Behavior**

Success: no issues found in 1 source file

**Your Environment**

mypy Playground  
- Mypy version used: mypy latest 1.17.1
- Mypy command-line flags: --strict
- Python version used: 3.12

**Comment by brianschubert** [comment]
Thanks for the report! Here's a reproducer that doesn't depend on `iter`'s signature in typeshed:
```python
from typing import Any, Protocol, Iterator

class _GetItemStrIterable(Protocol):
    def __getitem__(self, index: int, /) -> str: ...

def my_iter(x: _GetItemStrIterable, /) -> Iterator[str]: raise NotImplementedError

d: dict[int, str]
it: Iterator[str] = reveal_type(my_iter(d))  # N: Revealed type is "typing.Iterator[builtins.str]"
```

I think this is what's going on:

1. Mypy tries to use the type context (`Iterator[str]`) to solve for the type variables / overloads of `iter`
2. In typeshed, `iter` has [an overload](https://github.com/python/typeshed/blob/85a787bba3057f1df44188b4c815069d2d931557/stdlib/builtins.pyi#L1529-L1530) that allows it to accept objects that follow the "old-style" iteration protocol. (This protocol only requires a `__getitem__` implementation that accepts integers ascending from 0)
3. Using the above, mypy resolves the signature of `iter` into the signature of `my_iter` in the snippet above. This signature returns an `Iterator[str]` and accepts any object that has a `__getitem__` implementation that takes integers and returns strings.
4. The argument being passed to `iter` is compatible with this signature, since `dict[int, str]` has a `__getitem__` implementation  that takes integers and returns strings.
5. Since the argument types are compatible with the inferred signature, mypy is happy and concludes that `iter(d)` has type `Iterator[str]`

Strictly speaking, I think mypy is working correctly based on the information available to it in typeshed. The issue is that the signature of `iter` in typeshed doesn't quite reflect the runtime behavior, since the "old-style" iteration protocol is only applicable if an object doesn't have an `__iter__` method. Ideally, this would somehow be captured in the typeshed signature, but I'm not sure how. We can't, for example, write a protocol that only matches object which *do not* implement `__iter__`.

**Comment by sterliakov** [comment]
Yet another case where `NoMatch`/`@no_match`/`@type_error` for overloads ([discourse](https://discuss.python.org/t/specifying-invalid-signatures-in-overloads/75405/14)) would immensely help. A nomatch overload for `__iter__(self) -> Any` protocol before old-style protocol would have prevented this. I agree there's no mypy bug here: the typeshed defn is imprecise, and probably can't be corrected within current type system.

**Comment by didi64** [comment]
@brianschubert
Thanks for your very helpful explanation!
I understand now that this is not a bug, as iter(d) is assignable to Iterator[str], due to the second
<a href="https://github.com/python/typeshed/blob/85a787bba3057f1df44188b4c815069d2d931557/stdlib/builtins.pyi#L1529-L1530">overload</a> and that "old-style" iteration protocol.
